### PR TITLE
Merge `feature/protocol-state-kvstore` to `master`

### PR DESCRIPTION
**Pull request by jordanschalm**
Merges the `Protocol State KV Store` feature branch to `master`. 

There were many conflicts reported by Git. The vast majority were changes on `master` that had no conflicting change on the feature branch. There were some conflicts in integration test suite (some changed method names and re-organization) and in tools and API methods for retrieving a protocol snapshot. 

The full list of conflicting files is in the description of the merge commit: https://github.com/onflow/flow-go/pull/5840/commits/81b938ed4461fdc6f428890f480f96b7b8367959.

Changes made after the merge commit are in this [diff](https://github.com/onflow/flow-go/pull/5840/files/81b938ed4461fdc6f428890f480f96b7b8367959..9ccaee0fbad68db481e7c8abcf54daae003af55c)

## Outstanding TODOs
- [x] Resolve `core-contracts` dependency https://github.com/onflow/flow-go/pull/5840#discussion_r1589483631

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -50,8 +50,8 @@ require (
 	github.com/onflow/cadence v0.42.10
 	github.com/onflow/crypto v0.25.1
 	github.com/onflow/flow v0.3.4
-	github.com/onflow/flow-core-contracts/lib/go/contracts v0.15.1
-	github.com/onflow/flow-core-contracts/lib/go/templates v0.15.1
+	github.com/onflow/flow-core-contracts/lib/go/contracts v0.15.2-0.20240318135301-f59efe400a8b
+	github.com/onflow/flow-core-contracts/lib/go/templates v0.15.2-0.20240318135301-f59efe400a8b
```
Comment (turbolent) [inline go.mod:54]: Have these changes to the core contracts on master already been ported to the stable cadence feature branch (in the core contracts repo)?

File: cmd/bootstrap/cmd/finalize_test.go [hunk 2]
```diff
@@ -92,6 +93,7 @@ func TestFinalize_HappyPath(t *testing.T) {
 		log = log.Hook(hook)
 
 		finalize(nil, nil)
+		fmt.Println(hook.logs.String())
```
Comment (turbolent) [inline cmd/bootstrap/cmd/finalize_test.go:96]: Debug line?

File: cmd/util/ledger/migrations/utils.go [hunk 3]
```diff
@@ -40,26 +57,19 @@ func checkStorageHealth(
 		}
 	}
 
-	for _, domain := range allStorageMapDomains {
+	for _, domain := range domains {
```
Comment (turbolent) [inline cmd/util/ledger/migrations/utils.go:60]: Looks like a wrong conflict resolution? `domains` got renamed to `allStorageMapDomains` and moved above. `domains` below should be removed

File: fvm/evm/testutils/contracts/test_bytes.hex
Comment (turbolent) [inline fvm/evm/testutils/contracts/test_bytes.hex:1]: Is the only diff here the newline?

File: integration/benchmark/server/flow-go
Comment (turbolent) [inline integration/benchmark/server/flow-go:1]: Is adding this submodule intentional?

File: model/convert/fixtures_test.go
Comment (turbolent) [inline model/convert/fixtures_test.go:1]: Was the removal of this whole test file intentional?

**Review by peterargue — APPROVED** [review]
Added one comment about an unneeded access change. otherwise this looks good from the AN perspective.

**Review by turbolent — COMMENTED** [review]
Nice!

Did a cursory review of changes interacting with the Cadence 1.0 and migration code, and left one question regarding the core contract dependencies.

File: fvm/evm/testutils/contracts/test_bytes.hex
Comment (jordanschalm) [reply fvm/evm/testutils/contracts/test_bytes.hex:1]: Yes, it's only the newline.

File: integration/benchmark/server/flow-go
Comment (jordanschalm) [reply integration/benchmark/server/flow-go:1]: No 🤔. Removed the two submodule files.

File: model/convert/fixtures_test.go
Comment (jordanschalm) [reply model/convert/fixtures_test.go:1]: Yes, these fixtures had already been replaced by [this file](https://github.com/onflow/flow-go/blob/76cd5d469534e7c7e0e6904f28206df5b815d806/utils/unittest/service_events_fixtures.go) and were unused.

File: cmd/util/ledger/migrations/utils.go (hunk 3 above)
Comment (jordanschalm) [reply cmd/util/ledger/migrations/utils.go]: Fixed in https://github.com/onflow/flow-go/pull/5840/commits/d00bb5351c98dfc8db5da47a08475f9ac4d8d635

File: go.mod (hunk 1 above)
Comment (jordanschalm) [reply go.mod:54]: There is a stable Cadence version (https://github.com/onflow/flow-core-contracts/pull/419), but this is pointing to the pre-stable Cadence version. Both are still on feature branches. 

Thanks for pointing this out. I was hoping to implement https://github.com/onflow/flow-go/issues/5828 (allows simplifying the pending smart contract change) before merging the pinned branches. But we should pin a merged and tagged version of `core-contracts` before merging to `master`. 

We're not planning to deploy the smart contract change any time soon (hence why it hasn't been merged) and there is only one integration test which depends on the smart contract change. I might just disable that test prior to merging, and go back to the same `core-contracts` version from `master`.

File: engine/common/rpc/convert/transactions.go [hunk 4]
```diff
@@ -6,6 +6,14 @@ import (
 	"github.com/onflow/flow-go/model/flow"
 )
 
+// TransactionSubscribeInfo represents information about a subscribed transaction.
+// It contains the ID of the transaction, its status, and the index of the associated message.
+type TransactionSubscribeInfo struct {
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply engine/common/rpc/convert/transactions.go]: Removed in https://github.com/onflow/flow-go/pull/5840/commits/cf9719f2515020e37210d9cd600be809992734af

**Review by AlexHentschel — APPROVED** [review]
thanks for taking on the merge. Didn't find anything.

File: go.mod (hunk 1 above)
Comment (jordanschalm) [reply go.mod]: Disabled the test and reverted to `v0.15.1` in https://github.com/onflow/flow-go/pull/5840/commits/06978a492d9fc7b4e67087120af0979e3396b894.

**Review by turbolent — COMMENTED** [review]
Nice! 👏

**Pull request merged**